ERCS/eip-7772.md

```diff
@@ -0,0 +1,265 @@
+  - Parameters: 
+    - `bytes calldata data`: Additional data that can be passed to the dApp to help the dApp determine the amount of gas it is willing to sponsor.
+  - Visibility: External
+  - Usage: This function is called by the Paymaster during the validatePaymasterUserOp stage to set the gas that the dApp is willing to sponsor for the User Operation.
+
+### DApp Example
+```solidity
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import "./interfaces/IGasSponsor.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+contract Dapp is IGasSponsor {
+
+    constructor(....) {
+        ....
+    }
+
+    function swap(....) public {
+        
+        ....
+        
+        bytes32 userAccount = bytes32(bytes20(msg.sender));
+        assembly {
+            tstore(userAccount, 1000000000000000)
+        }
+    }
+    
+    function authoriseGas(
+        bytes calldata data
+    ) external override {
+        try {
+            // Any custom logic to decode the data
+            (
+                address someAddress,
+                bytes memory someData,
+                bytes memory someOtherData
+            ) = abi.decode(data, (address, bytes, bytes));
+            
+            if (...) {
+                assembly {
+                    tstore(msg.sender, 1000000000000000)
+                }
+            }
+        } catch {
+            // Handle the case where data doesn't conform to the expected format
+            // Leave blank to not authorise any amount
+        }    
+    }
+
+    function sponsorGas(
+        uint256 amount
+    ) external override {
+        uint256 allowance;
+        assembly {
+            allowance := tload(msg.sender)
+            tstore(msg.sender, 0) // Reset the value to zero to prevent attacks
+        }
+        payable(msg.sender).transfer(Math.min(allowance, amount));
+    }
+}
+```
+
+### TokenPaymaster Example
+The following is an example with TokenPaymaster. The highlighted portion SHOULD be included in the postOp regardless of the type of Paymaster.
+
+```solidity
+function _validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 maxCost
+    ) external override returns (bytes memory context, uint256 validationData) {
+
+        (
+            uint48 validUntil,
+            uint48 validAfter,
+            bytes calldata signature
+            bytes calldata sponsors
+        ) = parsePaymasterAndData(userOp.paymasterAndData);
+        
+        ...
+                 
+        // Iterate over the sponsors to collect gas
+        for (uint256 i = 0; i < sponsors.length; i += 20) {
+            address gasSponsor = address(bytes20(sponsors[i:i + 20]));
+
+            try IGasSponsor(gasSponsor).authoriseGas(userOp.paymasterAndData) {
+                emit GasAuthorised(gasSponsor);
+            } catch (bytes memory reason) {
```

Comment:
The paymaster trusts the sponsor codes. Remember, it is the user who controls which sponsors are added to the list, and the paymaster just obey that list.
e.g. a sponsor can waste gas, forcing the paymaster to pay for it, and even force revert (by wasting gas, and/or by `assembly { revert(0,100000) }`, causing the paymaster to waste a lot of gas to copy and emit this "error reason".